### ISSUE TITLE
Size of TFRecord is much more larger than CSV format

### ISSUE THREAD

**Issue by wenruij** [tensorflow/tensorflow#13577]
### Testing Data:
 `adult.data` in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/examples/learn/wide_n_deep_tutorial.py
### Problem:
To use a `tf.contrib.data.TFRecordDataset`, i tried to convert `adult.data` in CSV into a TFRecord, but i just found that the **TFRecord** after converted is about **12MB**, while the original **CSV** is only about **3MB**, oops, **why the storage efficiency for TFRecord is so poor？**

### Source code for converting CSV to TFRecord:
```
def _int64_feature(value):
    return tf.train.Feature(int64_list=tf.train.Int64List(value=value))

def _bytes_feature(value):
    return tf.train.Feature(bytes_list=tf.train.BytesList(value=value))

def csv2proto(
        data_source,
        all_cols,
        categorical_cols=[],
        continuous_cols=[],
        multi_values_cols=[],
        inner_delimiter=';'):
    if not os.path.isfile(data_source):
        raise ValueError('data file passed do not exist or not a file')

    file_name = os.path.splitext(data_source)[0] + '.tfrecords'
    writer = tf.python_io.TFRecordWriter(file_name)
    with open(data_source) as f:
        reader = csv.DictReader(f, fieldnames=all_cols)
        for row in reader:
            feature = dict()
            for col in categorical_cols:
                feature.update({col: _bytes_feature([row[col]])})
            for col in continuous_cols:
                feature.update({col: _int64_feature([int(row[col])])})
            for col in multi_values_cols:
                feature.update({col: _bytes_feature(row[col].split(inner_delimiter))})

            example = tf.train.Example(features=tf.train.Features(feature=feature))
            writer.write(example.SerializeToString())

        writer.close()
```

### System information:
- **OS Platform**:  Mac OS X 10.12.5
- **TensorFlow installed from (source or binary)**: pip install
- **TensorFlow version (use command below)**: 1.3.0
- **Python version**:  2.7
- **Bazel version (if compiling from source)**: None
- **CUDA/cuDNN version**: None
- **GPU model and memory**: None

**Comment by reedwm** [comment]
Please include the code that calls `csv2proto`, so I have a self-contained example I can run to generate the TFRecord.

**Comment by wenruij** [comment]
@reedwm pls check the full code as below,

```
import csv
import os
import tempfile

from six.moves import urllib
from datetime import datetime

import tensorflow as tf


MB_SIZE = 1024.0*1024.0
CSV_FORMAT = 'adult.data'

categorical_columns = [
    "workclass", "education", "marital_status", "occupation", "relationship",
    "race", "gender", "native_country", "label"
]
continuous_columns = [
    "age", "fnlwgt", "education_num", "capital_gain", "capital_loss",
    "hours_per_week"
]
all_columns = [
    "age", "workclass", "fnlwgt", "education", "education_num",
    "marital_status", "occupation", "relationship", "race", "gender",
    "capital_gain", "capital_loss", "hours_per_week", "native_country", "label"
]


def _int64_feature(value):
    return tf.train.Feature(int64_list=tf.train.Int64List(value=value))


def _bytes_feature(value):
    return tf.train.Feature(bytes_list=tf.train.BytesList(value=value))


def csv2proto(
        data_source,
        all_cols,
        categorical_cols=[],
        continuous_cols=[],
        multi_values_cols=[],
        inner_delimiter=';'
):
    if not os.path.isfile(data_source):
        raise ValueError('data file passed do not exist or not a file')

    csv_size = os.path.getsize(data_source)/MB_SIZE
    file_name = os.path.splitext(data_source)[0] + '.tfrecords'
    writer = tf.python_io.TFRecordWriter(file_name)
    start = datetime.now()
    with open(data_source) as f:
        reader = csv.DictReader(f, fieldnames=all_cols)
        for row in reader:
            feature = dict()
            for col in categorical_cols:
                feature.update({col: _bytes_feature([row[col]])})
            for col in continuous_cols:
                feature.update({col: _int64_feature([int(row[col])])})
            for col in multi_values_cols:
                feature.update({col: _bytes_feature(row[col].split(inner_delimiter))})

            example = tf.train.Example(features=tf.train.Features(feature=feature))
            writer.write(example.SerializeToString())

        writer.close()
        end = datetime.now()
        proto_size = os.path.getsize(file_name)/MB_SIZE

        convert_logs = "\n| Time Elapsed: %d" % (end-start).seconds + " s |"
        convert_logs += "\n| csv size: " + "%.2f" % csv_size + " MB |"
        convert_logs += "\n| proto size: " + "%.2f" % proto_size + " MB |\n"

        print convert_logs


def main():
    urllib.request.urlretrieve(
        "https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data",
        CSV_FORMAT)
    csv2proto(data_source=CSV_FORMAT,
              all_cols=all_columns,
              categorical_cols=categorical_columns,
              continuous_cols=continuous_columns)


if __name__ == "__main__":
    main()
```

Logs of my side(Mac OS X 10.12.5) is:
```
| Time Elapsed: 89 s |
| csv size: 3.79 MB |
| proto size: 12.68 MB |
```

**Comment by reedwm** [comment]
@davidsoergel, do you know why `tf.train.Example`s are significantly larger than the equivalent row of a CSV file?

**Comment by Moondra** [comment]
I'm running into a similar situation using images. 

I uploaded my code here:

https://github.com/moondra2017/Testing_Cat_classifier_Tensorflow/blob/master/Creating_tf_records_of_small_set_fishes.py

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by wenruij** [comment]
Its a good way to serialize a libsvm into tf.Example protos in this way:
```
def libsvm2proto(data_source, target_dir, delimiter='\t'):
    """
    a single file should not contain lines more than 1000,000
    :param data_source: libsvm file path
    :param target_dir: dir to storage the serialize proto file
    :param delimiter: delimiter for csv reader
    :return:
    """
    if not os.path.isfile(data_source):
        raise ValueError('data file passed do not exist or not a file')

    file_name = os.path.join(target_dir, os.path.splitext(
        os.path.basename(data_source))[0] + '.tfrecords')
    writer = tf.python_io.TFRecordWriter(file_name)
    start = datetime.now()
    with open(data_source, 'rb') as rf:
        f_reader = csv.reader(rf, delimiter=delimiter, quotechar='|')
        for row in f_reader:
            feature = dict()
            indexes = []
            values = []
            feature.update({'label': _float_feature([float(row[0])])})
            for e in row[1:]:
                index, value = e.split(':')
                indexes.append(int(index))
                values.append(float(value))
                feature.update({'index': _int64_feature(indexes)})
                feature.update({'value': _float_feature(values)})

            example = tf.train.Example(features=tf.train.Features(feature=feature))
            writer.write(example.SerializeToString())

        writer.close()
        end = datetime.now()

        print("- consumed time: %ds for %s" % ((end-start).seconds, data_source))
```
The size of TFRecord after converted is just half of the source libsvm 👍 .  That's what we really expect.
